subsys/gazell/gzll_glue.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/nrf_clock_control.h>
+
+#include <nrf_gzll_glue.h>
+#include <gzll_glue.h>
+#include <nrfx_ppi.h>
+
+
+#if defined(CONFIG_GAZELL_TIMER0)
+#define GAZELL_TIMER_IRQN           TIMER0_IRQn
+NRF_TIMER_Type * const nrf_gzll_timer = NRF_TIMER0;
+#elif defined(CONFIG_GAZELL_TIMER1)
+#define GAZELL_TIMER_IRQN           TIMER1_IRQn
+NRF_TIMER_Type * const nrf_gzll_timer = NRF_TIMER1;
+#elif defined(CONFIG_GAZELL_TIMER2)
+#define GAZELL_TIMER_IRQN           TIMER2_IRQn
+NRF_TIMER_Type * const nrf_gzll_timer = NRF_TIMER2;
+#elif defined(CONFIG_GAZELL_TIMER3)
+#define GAZELL_TIMER_IRQN           TIMER3_IRQn
+NRF_TIMER_Type * const nrf_gzll_timer = NRF_TIMER3;
+#elif defined(CONFIG_GAZELL_TIMER4)
+#define GAZELL_TIMER_IRQN           TIMER4_IRQn
+NRF_TIMER_Type * const nrf_gzll_timer = NRF_TIMER4;
+#else
+#error "Gazell timer is undefined."
+#endif
+IRQn_Type        const nrf_gzll_timer_irqn = GAZELL_TIMER_IRQN;
+
+#if defined(CONFIG_GAZELL_SWI0)
+#define GAZELL_SWI_IRQN             SWI0_IRQn
+#elif defined(CONFIG_GAZELL_SWI1)
+#define GAZELL_SWI_IRQN             SWI1_IRQn
+#elif defined(CONFIG_GAZELL_SWI2)
+#define GAZELL_SWI_IRQN             SWI2_IRQn
+#elif defined(CONFIG_GAZELL_SWI3)
+#define GAZELL_SWI_IRQN             SWI3_IRQn
+#elif defined(CONFIG_GAZELL_SWI4)
+#define GAZELL_SWI_IRQN             SWI4_IRQn
+#elif defined(CONFIG_GAZELL_SWI5)
+#define GAZELL_SWI_IRQN             SWI5_IRQn
+#else
+#error "Gazell software interrupt is undefined."
+#endif
+IRQn_Type        const nrf_gzll_swi_irqn   = GAZELL_SWI_IRQN;
+
+__IOM uint32_t *nrf_gzll_ppi_eep0;
+__IOM uint32_t *nrf_gzll_ppi_tep0;
+__IOM uint32_t *nrf_gzll_ppi_eep1;
+__IOM uint32_t *nrf_gzll_ppi_tep1;
+__IOM uint32_t *nrf_gzll_ppi_eep2;
+__IOM uint32_t *nrf_gzll_ppi_tep2;
+
+uint32_t nrf_gzll_ppi_chen_msk_0_and_1;
+uint32_t nrf_gzll_ppi_chen_msk_2;
+
+
+ISR_DIRECT_DECLARE(gazell_radio_irq_handler)
+{
+	nrf_gzll_radio_irq_handler();
+
+	return 0;
```

Comment:
Consider calling `ISR_DIRECT_PM()` and returning 1 if Gazell event handler in the application is expected to trigger a context switch (e.g. unblock some thread etc.). See https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/esb/esb.c#L996

Reply:
The current code should be more efficient.

- When the Gazell radio IRQ handler has a packet transaction completed, it will trigger a SWI. The radio IRQ handler can return a 0 here. The SWI handler, gazell_swi_irq_handler, goes through the normal context switching procedure.
- When the Gazell radio IRQ handler has no transaction done, it can return a 0. No context switch is needed.
- Always returning 1 triggers lot of context switches without signals waiting by the application.
- The ISR_DIRECT_PM() is traced. It looks waking up the Zephyr kernel.
`ISR_DIRECT_PM() --> __arch_isr_direct_pm() --> z_pm_save_idle_exit() --> pm_system_resume() --> sys_clock_idle_exit()`
- The Gazell host switches radio frequency channel every millisecond. Calling ISR_DIRECT_PM() means kernel exiting idle frequently.
- ISR_DIRECT_PM() is marked optional in Zephyr documentation.

Reply:
Thanks for explaining @leewkb4567! Right, seems like the Radio IRQ should not enable context switching, as there are SWI which allow context switch.